client/blocks/payment-request/use-checkout-subscriptions.js

```diff
@@ -0,0 +1,241 @@
+	}, [
+		paymentRequestEventHandlers,
+		billing,
+		shippingData,
+		paymentRequestType,
+	] );
+
+	// subscribe to events.
+	useEffect( () => {
+		const onShippingRatesEvent = ( shippingRates ) => {
+			const handlers = eventHandlers.current;
+			const billingData = currentBilling.current;
+			if ( handlers.shippingAddressChange && isProcessing ) {
+				handlers.shippingAddressChange.updateWith( {
+					status: 'success',
+					shippingOptions: normalizeShippingOptions( shippingRates ),
+					total: getTotalPaymentItem( billingData.cartTotal ),
+					displayItems: normalizeLineItems(
+						billingData.cartTotalItems
+					),
+				} );
+				clearPaymentRequestEventHandler( 'shippingAddressChange' );
+			}
+		};
+		const onShippingRatesEventFail = ( currentErrorStatus ) => {
+			const handlers = eventHandlers.current;
+			if ( handlers.shippingAddressChange && isProcessing ) {
+				handlers.shippingAddressChange.updateWith( {
+					status: currentErrorStatus.hasInvalidAddress
+						? 'invalid_shipping_address'
+						: 'fail',
+					shippingOptions: [],
+				} );
+			}
+			clearPaymentRequestEventHandler( 'shippingAddressChange' );
+		};
+		const onShippingSelectedRate = ( forSuccess = true ) => () => {
+			const handlers = eventHandlers.current;
+			const shipping = currentShipping.current;
+			const billingData = currentBilling.current;
+			if (
+				handlers.shippingOptionChange &&
+				! shipping.isSelectingRate &&
+				isProcessing
+			) {
+				const updateObject = forSuccess
+					? {
+							status: 'success',
+							total: getTotalPaymentItem( billingData.cartTotal ),
+							displayItems: normalizeLineItems(
+								billingData.cartTotalItems
+							),
+					  }
+					: {
+							status: 'fail',
+					  };
+				handlers.shippingOptionChange.updateWith( updateObject );
+				clearPaymentRequestEventHandler( 'shippingOptionChange' );
+			}
+		};
+		const onProcessingPayment = () => {
+			const handlers = eventHandlers.current;
+			if ( handlers.sourceEvent && isProcessing ) {
+				const response = {
+					type: responseTypes.SUCCESS,
+					meta: {
+						billingData: getBillingData( handlers.sourceEvent ),
+						paymentMethodData: getPaymentMethodData(
+							handlers.sourceEvent,
+							currentPaymentRequestType.current
+						),
+						shippingData: getShippingData( handlers.sourceEvent ),
+					},
+				};
+				return response;
+			}
+			return { type: responseTypes.SUCCESS };
+		};
+		const onCheckoutComplete = ( checkoutResponse ) => {
+			const handlers = eventHandlers.current;
+			let response = { type: responseTypes.SUCCESS };
+			if ( handlers.sourceEvent && isProcessing ) {
+				const {
+					paymentStatus,
+					paymentDetails,
+				} = checkoutResponse.processingResponse;
+				if ( paymentStatus === responseTypes.SUCCESS ) {
+					completePayment( handlers.sourceEvent );
+				}
+				if (
+					paymentStatus === responseTypes.ERROR ||
+					paymentStatus === responseTypes.FAIL
+				) {
```

Comment:
Let's use the established [yoda condition WP standard](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#yoda-conditions) even though that's _technically_ only defined for PHP 🙂